sql/plan/indexed_join.go

```diff
@@ -0,0 +1,253 @@
+	})
+
+	if indexedTable == nil {
+		return nil, ErrNoIndexedTableAccess.New(ij.Right)
+	}
+
+	return indexedJoinRowIter(ctx, ij.Left, ij.Right, indexedTable, ij.primaryTableExpr, ij.Cond, ij.Index, ij.joinType)
+}
+
+func (ij *IndexedJoin) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(ij, len(children), 2)
+	}
+	return NewIndexedJoin(children[0], children[1], ij.joinType, ij.Cond, ij.primaryTableExpr, ij.Index), nil
+}
+
+func indexedJoinRowIter(ctx *sql.Context, left sql.Node, right sql.Node, indexAccess *IndexedTableAccess, primaryTableExpr sql.Expression, cond sql.Expression, index sql.Index, joinType JoinType) (sql.RowIter, error) {
+	var leftName, rightName string
+	if leftTable, ok := left.(sql.Nameable); ok {
+		leftName = leftTable.Name()
+	} else {
+		leftName = reflect.TypeOf(left).String()
+	}
+
+	if rightTable, ok := right.(sql.Nameable); ok {
+		rightName = rightTable.Name()
+	} else {
+		rightName = reflect.TypeOf(right).String()
+	}
+
+	span, ctx := ctx.Span("plan.indexedJoin", opentracing.Tags{
+		"left":  leftName,
+		"right": rightName,
+	})
+
+	l, err := left.RowIter(ctx)
+	if err != nil {
+		span.Finish()
+		return nil, err
+	}
+	return sql.NewSpanIter(span, &indexedJoinIter{
+		primary:              l,
+		secondaryProvider:    right,
+		secondaryIndexAccess: indexAccess,
+		ctx:                  ctx,
+		cond:                 cond,
+		primaryTableExpr:     primaryTableExpr,
+		index:                index,
+		joinType:             joinType,
+		rowSize:              len(left.Schema()) + len(right.Schema()),
+	}), nil
+}
+
+// indexedJoinIter is an iterator that iterates over every row in the primary table and performs an index lookup in
+// the secondary table for each value
+type indexedJoinIter struct {
+	primary              sql.RowIter
+	primaryRow           sql.Row
+	index                sql.Index
+	secondaryIndexAccess *IndexedTableAccess
+	secondaryProvider    sql.Node
+	secondary            sql.RowIter
+	primaryTableExpr     sql.Expression
+	cond                 sql.Expression
+	joinType             JoinType
+
+	ctx        *sql.Context
+	foundMatch bool
+	rowSize    int
+}
+
+func (i *indexedJoinIter) loadPrimary() error {
+	if i.primaryRow == nil {
+		r, err := i.primary.Next()
+		if err != nil {
+			return err
+		}
+
+		i.primaryRow = r
+		i.foundMatch = false
+	}
+
+	return nil
+}
+
+func (i *indexedJoinIter) loadSecondary() (sql.Row, error) {
+	if i.secondary == nil {
+		// evaluate the primary row against the primary table expression to get the secondary table lookup key
+		key, err := i.primaryTableExpr.Eval(i.ctx, i.primaryRow)
+		if err != nil {
+			return nil, err
+		}
+
+		lookup, err := i.index.Get(key)
+		if err != nil {
+			return nil, err
+		}
+
+		err = i.secondaryIndexAccess.SetIndexLookup(i.ctx, lookup)
+		if err != nil {
+			return nil, err
+		}
+
+		span, ctx := i.ctx.Span("plan.IndexedJoin indexed lookup")
```

Comment:
Should technically be a child Span?

Reply:
It is, that's what ctx.Span does. Unless it has a bug.